src/storage-client/src/storage_collections.rs

```diff
@@ -519,10 +520,71 @@ where
         since: Option<&Antichain<T>>,
         relation_desc: RelationDesc,
         persist_client: &PersistClient,
-    ) -> (
-        WriteHandle<SourceData, (), T, Diff>,
-        SinceHandle<SourceData, (), T, Diff, PersistEpoch>,
-    ) {
+    ) -> (WriteHandle<SourceData, (), T, Diff>, SinceHandleWrapper<T>) {
+        let write_handle = self
+            .open_write_handle(id, shard, relation_desc, persist_client)
+            .await;
+        let since_handle = self
+            .open_critical_handle(id, shard, since, persist_client)
+            .await;
+
+        let since_handle = SinceHandleWrapper::Critical(since_handle);
+
+        (write_handle, since_handle)
+    }
+
+    /// Opens a write handle for the given `shard`.
+    async fn open_write_handle(
+        &self,
+        id: &GlobalId,
+        shard: ShardId,
+        relation_desc: RelationDesc,
+        persist_client: &PersistClient,
+    ) -> WriteHandle<SourceData, (), T, Diff> {
+        let diagnostics = Diagnostics {
+            shard_name: id.to_string(),
+            handle_purpose: format!("controller data for {}", id),
+        };
+
+        let mut write = persist_client
+            .open_writer(
+                shard,
+                Arc::new(relation_desc),
+                Arc::new(UnitSchema),
+                diagnostics.clone(),
+            )
+            .await
+            .expect("invalid persist usage");
+
+        // N.B.
+        // Fetch the most recent upper for the write handle. Otherwise, this may
+        // be behind the since of the since handle. Its vital this happens AFTER
+        // we create the since handle as it needs to be linearized with that
+        // operation. It may be true that creating the write handle after the
+        // since handle already ensures this, but we do this out of an abundance
+        // of caution.
```

Comment:
The comment here says that it's vital to do the below AFTER we have created the since handle, but there is nothing in this method that ensures this. In fact, the caller of this method calls it before creating the since handle. Is this an issue or is the comment misleading (I don't really understand it)?

Reply:
I myself am not 100% sure that we need this. I'll change this PR to have both this comment and the call to `fetch_recent_upper` in `open_data_handles` again. That way we at least ensure we have the previous behavior and the comment makes sense again.